doc/tutorials/04_errors.md

```diff
@@ -0,0 +1,219 @@
+# 4. Error handling
+
+Many things can go wrong: that's why error handling is often the centerpiece of software libraries. We got a glimpse of 
+one of Tapir's components when it comes to error handling when we discussed 
+[adding OpenAPI documentation](02_openapi_docs.md). In this tutorial, we'll investigate Tapir's approach to error 
+handling in more detail.
+
+Errors might be divided into "expected" errors: that is ones that we know how to handle, for which we have designed a 
+specific response. These errors are most often caused by invalid input from the user (that is, invalid data that's part 
+of the HTTP request). For such requests, we should return responses with error codes between 400 and 499, which are 
+designated in the HTTP specification as "client errors".
+
+On the other hand, there are "unexpected errors", that we didn't foresee. When they occur, they signal a bug in
+the server's logic. When this happens, we should respond with a status code between 500 and 599, and log the error
```

Comment:
**suggestion:** sometimes we return 503 when there are performance issues of our backend/infrastructure or a dependency, so it's not necessarily caused by a "bug", right?

Reply:
ah yes, right. I'll mention this as well, thanks